sycl/source/detail/kernel_bundle_impl.hpp

```diff
@@ -0,0 +1,175 @@
+#include <vector>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+
+// The class is an impl counterpart of the sycl::kernel_bundle.
+// It provides an access and utilities to manage set of sycl::device_images
+// objects.
+class kernel_bundle_impl {
+
+public:
+  kernel_bundle_impl(const context &Ctx, const std::vector<device> &Devs,
+                     bundle_state State)
+      : MContext(Ctx), MDevices(Devs) {
+
+    MDeviceImages = detail::ProgramManager::getInstance().getSYCLDeviceImages(
+        Ctx, Devs, State);
+  }
+
+  kernel_bundle_impl(const context &Ctx, const std::vector<device> &Devs,
+                     const std::vector<kernel_id> &KernelIDs,
+                     bundle_state State)
+      : kernel_bundle_impl(Ctx, Devs, State) {
+
+    // Filter out images that have no kernel_ids specified
+    auto It = std::remove_if(MDeviceImages.begin(), MDeviceImages.end(),
+                             [&KernelIDs](const device_image_plain &Image) {
+                               return std::none_of(
+                                   KernelIDs.begin(), KernelIDs.end(),
+                                   [&Image](const sycl::kernel_id &KernelID) {
+                                     return Image.has_kernel(KernelID);
+                                   });
+                             });
+    MDeviceImages.erase(It, MDeviceImages.end());
+  }
+
+  kernel_bundle_impl(const context &Ctx, const std::vector<device> &Devs,
+                     const DevImgSelectorImpl &Selector, bundle_state State)
+      : kernel_bundle_impl(Ctx, Devs, State) {
+
+    // Filter out images that are rejected by Selector
+    auto It = std::remove_if(MDeviceImages.begin(), MDeviceImages.end(),
+                             [&Selector](const device_image_plain &Image) {
+                               return !Selector(getSyclObjImpl(Image));
+                             });
+    MDeviceImages.erase(It, MDeviceImages.end());
+  }
+
+  bool empty() const noexcept { return MDeviceImages.empty(); }
+
+  backend get_backend() const noexcept {
+    return MContext.get_platform().get_backend();
+  }
+
+  context get_context() const noexcept { return MContext; }
+
+  std::vector<device> get_devices() const noexcept { return MDevices; }
+
+  std::vector<kernel_id> get_kernel_ids() const {
+    // Collect kernel ids from all device images, then remove duplicates
+
+    std::vector<kernel_id> Result;
+    for (const device_image_plain &DeviceImage : MDeviceImages) {
+      const std::vector<kernel_id> &KernelIDs =
+          getSyclObjImpl(DeviceImage)->get_kernel_ids();
+
+      Result.insert(Result.end(), KernelIDs.begin(), KernelIDs.end());
+    }
+    std::sort(Result.begin(), Result.end(),
+              [](const kernel_id &LHS, const kernel_id &RHS) {
+                return detail::getSyclObjImpl(LHS) <
+                       detail::getSyclObjImpl(RHS);
+              });
+    std::unique(Result.begin(), Result.end());
```

Comment:
The code seems complex.
Would it make sense to use some `std::map` instead to solve the unique and sort aspect while removing the need to do some binary searches with `std::lower_bound` and such?

Reply:
If we used `std::map` here we would need to do a "repacking" to `std::vector` at the end which looks like unnecessary overhead.

Reply:
This is a good point.
Probably the next SYCL version will not impose `std::vector` but will just use C++20 `std::ranges` which might influence the design later. We will see...